### PR TITLE
[SPARK-31377][SQL][TEST] Added unit tests to 'number of output rows metric' for some joins in SQLMetricSuite

### PR DESCRIPTION
**Pull request by sririshindra**
### What changes were proposed in this pull request?
Add unit tests to the 'number of output rows metric' for some join types in the SQLMetricSuite. A list of unit tests added are as follows.
- ShuffledHashJoin: leftOuter, RightOuter, LeftAnti, LeftSemi
- BroadcastNestedLoopJoin: RightOuter
- BroadcastHashJoin: LeftAnti

### Why are the changes needed?
For some combinations of JoinType and Join algorithm there is no test coverage for the 'number of output rows' metric.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
I added debug statements in the code to ensure the correct combination if JoinType and Join algorithms are triggered.
I further used Intellij debugger to test the same.

### CODE REVIEW

**Comment by maropu** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 1]
```diff
@@ -341,6 +341,62 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
     }
   }
 
+  test("ShuffledHashJoin(outer) metrics") {
+    withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "40",
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:345]: `40` -> `-1`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 2]
```diff
@@ -341,6 +341,62 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
+      SQLConf.SHUFFLE_PARTITIONS.key -> "2",
+      SQLConf.PREFER_SORTMERGEJOIN.key -> "false") {
+      val df1 = Seq((1, "1"), (2, "2")).toDF("key", "value")
+      val df2 = (1 to 10).map(i => (i, i.toString)).toSeq.toDF("key2", "value")
+
+      Seq(("right_outer", 0L, df1, df2, false), ("left_outer", 0L, df2, df1, false),
+        ("right_outer", 0L, df1, df2, true), ("left_outer", 0L, df2, df1, true))
+        .foreach { case (joinType, nodeId, df1, df2, enableWholeStage) =>
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:353]: nit: `df1` -> `leftDf` and `df2` -> `rightDf`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 3]
```diff
@@ -341,6 +341,62 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
     }
   }
 
+  test("ShuffledHashJoin(outer) metrics") {
+    withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "40",
+      SQLConf.SHUFFLE_PARTITIONS.key -> "2",
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:346]: We need to set this value for the test?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 4]
```diff
@@ -341,6 +341,62 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
+      SQLConf.PREFER_SORTMERGEJOIN.key -> "false") {
+      val df1 = Seq((1, "1"), (2, "2")).toDF("key", "value")
+      val df2 = (1 to 10).map(i => (i, i.toString)).toSeq.toDF("key2", "value")
+
+      Seq((1L, false), (2L, true)).foreach { case (nodeId, enableWholeStage) =>
+        val df = df2.join(df1.hint("shuffle_hash"), $"key" === $"key2", "left_semi")
+        testSparkPlanMetrics(df, 1, Map(
+          nodeId -> (("ShuffledHashJoin", Map(
+            "number of output rows" -> 2L)))),
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:393]: We need to split this join test into three parts? It seems the only metric value is different between them.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 1 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:345]: Setting it to -1 would trigger SortMergeJoin instead of ShuffledHashJoin based on the rules in Spark Strategies.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 2 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: fixed in the latest commit

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 3 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:346]: In my latest commit combined three different tests into one. Not setting this parameter will not trigger the correct join type based on the rules in the Spark Strategies file.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 4 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: fixed in the latest commit

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 1 above)
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:345]: How about using hint?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 3 above)
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:346]: ditto: https://github.com/apache/spark/pull/28330#discussion_r415434157

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 5]
```diff
@@ -547,9 +590,9 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
 
   test("SPARK-25602: SparkPlan.getByteArrayRdd should not consume the input when not necessary") {
     def checkFilterAndRangeMetrics(
-        df: DataFrame,
-        filterNumOutputs: Int,
-        rangeNumOutputs: Int): Unit = {
+                                    df: DataFrame,
+                                    filterNumOutputs: Int,
+                                    rangeNumOutputs: Int): Unit = {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:595]: plz revert this.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 6]
```diff
@@ -325,36 +325,61 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
-        nodeId -> (("BroadcastHashJoin", Map(
-          "number of output rows" -> numRows)))),
-        enableWholeStage
-      )
+        val df = df1.join(broadcast(df2), $"key" === $"key2", joinType)
+        testSparkPlanMetrics(df, 2, Map(
+          nodeId -> (("BroadcastHashJoin", Map(
+            "number of output rows" -> numRows)))),
+          enableWholeStage
+        )
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:382]: nit: wrong indents

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 5 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: I am sorry about this. Somehow intellj added the extras spaces. Fixed in latest commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 6 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Fixed in latest commit

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 1 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Fixed in the latest commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 3 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Fixed in the latest commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 7]
```diff
@@ -325,34 +325,57 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
+          testSparkPlanMetrics(df, 1, Map(
+            nodeId1 -> (("ShuffledHashJoin", Map(
+              "number of output rows" -> 2L))),
+            nodeId2 -> (("Exchange", Map(
+              "shuffle records written" -> 2L,
+              "records read" -> 2L))),
+            nodeId3 -> (("Exchange", Map(
+              "shuffle records written" -> 10L,
+              "records read" -> 10L)))),
+            enableWholeStage
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:338]: unnecessary changes?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 8]
```diff
@@ -325,34 +325,57 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
+          )
       }
     }
   }
 
+  test("ShuffledHashJoin(left,outer) metrics") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:344]: nit: `(left, outer)`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 9]
```diff
@@ -325,34 +325,57 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
+    withSQLConf(SQLConf.SHUFFLE_PARTITIONS.key -> "2",
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:345]: Still needs this setting, even though the hint used?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 10]
```diff
@@ -325,34 +325,57 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
-    Seq(("left_outer", 0L, 5L, false), ("right_outer", 0L, 6L, false),
-      ("left_outer", 1L, 5L, true), ("right_outer", 1L, 6L, true)).foreach {
-      case (joinType, nodeId, numRows, enableWholeStage) =>
+    Seq(("left_outer", 0L, 5L, false), ("right_outer", 0L, 6L, false), ("left_outer", 1L, 5L, true),
+      ("right_outer", 1L, 6L, true)).foreach { case (joinType, nodeId, numRows, enableWholeStage) =>
       val df = df1.join(broadcast(df2), $"key" === $"key2", joinType)
       testSparkPlanMetrics(df, 2, Map(
-        nodeId -> (("BroadcastHashJoin", Map(
-          "number of output rows" -> numRows)))),
+        nodeId -> (("BroadcastHashJoin", Map("number of output rows" -> numRows)))),
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:378]: unnecessary changes?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 11]
```diff
@@ -394,6 +420,21 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
     }
   }
 
+  test("BroadcastLeftAntiJoinHash metrics") {
+    val df1 = Seq((1, "1"), (2, "2")).toDF("key", "value")
+    val df2 = Seq((1, "1"), (2, "2"), (3, "3"), (4, "4")).toDF("key2", "value")
+    // Assume the execution plan is
+    // ... -> BroadcastHashJoin(nodeId = 1)
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:427]: Need this comment?  I think the code below is clear without this comment.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 10 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Fixed in latest commit

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 11 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Fixed in latest commit

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 9 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: You are right. I should have removed these in the last commit itself. Fixed in the latest commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 8 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: fixed in the latest commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 7 above)
Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:338]: I think this indentation is correct. I know it is just a small cosmetic change and probably doesn't need to be included in this PR. I will remove it if you think this should not be there.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 12]
```diff
@@ -365,16 +384,19 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
...
+          "testData2.a * testDataForJoin.a != testData2.a + testDataForJoin.a"
+        Seq((leftQuery, false), (rightQuery, false), (leftQuery, true), (rightQuery, true))
+          .foreach { case (query, enableWholeStage) =>
+            val df = spark.sql(query)
+            testSparkPlanMetrics(df, 2, Map(
+              0L -> (("BroadcastNestedLoopJoin", Map(
+                "number of output rows" -> 12L)))),
+              enableWholeStage
+            )
+          }
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:399]: nit: format(wrong indents)
```
        Seq(
          (leftQuery, false),
          (rightQuery, false),
          (leftQuery, true),
          (rightQuery, true)).foreach { case (query, enableWholeStage) =>
          val df = spark.sql(query)
          testSparkPlanMetrics(df, 2, Map(
            0L -> (("BroadcastNestedLoopJoin", Map(
              "number of output rows" -> 12L)))),
            enableWholeStage
          )
        }
```

Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:399]: I tried reformatting this piece of code using intellij reformat feature and that didn't change anything. Could you please let me know where the indentation is wrong so that I can fix it.

Comment (sririshindra) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Fixed in latest commit.

**Comment by sririshindra** [comment]
@maropu Could you please let me know if there are any other changes needed in this PR. If not could you merge this PR. Thank you.

**Comment by sririshindra** [comment]
@maropu Could you please take a look at this when you have a moment.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 13]
```diff
@@ -310,8 +310,8 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
 
   test("ShuffledHashJoin metrics") {
     withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "40",
-        SQLConf.SHUFFLE_PARTITIONS.key -> "2",
-        SQLConf.PREFER_SORTMERGEJOIN.key -> "false") {
+      SQLConf.SHUFFLE_PARTITIONS.key -> "2",
+      SQLConf.PREFER_SORTMERGEJOIN.key -> "false") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:314]: nit: plz avoid unnecesary changes where possible.

**Review by maropu — APPROVED** [review]
cc: @HyukjinKwon @dongjoon-hyun

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.0.
